pyscriptjs/src/main.ts

```diff
@@ -76,25 +96,38 @@ export class PyScriptApp {
 
         this.plugins.add(new StdioDirector(this._stdioMultiplexer));
         this.tagExecutionLock = createLock();
+        this.numPendingTags = 0;
+        this.scriptTagsPromise = new Promise(res => (this.resolvedScriptTags = res));
     }
 
     // Error handling logic: if during the execution we encounter an error
     // which is ultimate responsibility of the user (e.g.: syntax error in the
     // config, file not found in fetch, etc.), we can throw UserError(). It is
     // responsibility of main() to catch it and show it to the user in a
     // proper way (e.g. by using a banner at the top of the page).
-    main() {
+    async main() {
         try {
-            this._realMain();
+            await this._realMain();
         } catch (error) {
             this._handleUserErrorMaybe(error);
         }
     }
 
-    _handleUserErrorMaybe(error) {
-        if (error instanceof UserError) {
+    incrementNumPendingTags() {
+        this.numPendingTags += 1;
+    }
+
+    decrementNumPendingTags() {
+        this.numPendingTags -= 1;
+        if (this.numPendingTags === 0) {
+            this.resolvedScriptTags();
+        }
+    }
```

Comment:
done

Reply:
my 2ct: to be honest to me it looks like `numPendingTags` is an implementation detail (private field) that shouldn't be checked or handled at all outside `main.ts` ... the two methods should already do the right thing so that we don't need to check `numPendingTags > 0` and then again its value within `decrementNumPendingTags` ... we should just be sure that if `incrementPendingTags()` is called, `decremengPendingTags()` is also called at some point, removing both the `Num` part of these methods and avoiding leaking internals to consumers.

Reply:
>> we should have a sanity check
> done

no, this is the "wrong" way of doing a sanity check. A sanity check means that we *know* it will never be negative, but we are paranoid and we want to check it anyway in case there is a bug somewhere.
So, if the sanity check fails, we should stop immediately and raise an "internal error" exception.